### PR TITLE
Add async_rw_mutex based on senders

### PR DESCRIPTION
**Pull request by msimberg**
Adds an `async_rw_mutex` (name not finalized, but I think this describes it fairly well). Instead of directly locking a mutex, the async mutex returns senders that give access to a resource when scheduled. The `rw` part means that access can be read-only (shared, multiple concurrent readers allowed), or exclusively read-write.

This is based on an implementation used in DLA-Future where they protect access to tiles of a matrix in this fashion. That mechanism is based on futures/promises and custom for their matrix class. This is an attempt at a more generic version.

Contains #5274 so the diff is rather large.

### CODE REVIEW

File: libs/core/type_support/include/hpx/type_support/detail/with_result_of.hpp
Comment (hkaiser) [inline libs/core/type_support/include/hpx/type_support/detail/with_result_of.hpp:21]: nitpick: `typename` ?

File: libs/core/type_support/include/hpx/type_support/detail/with_result_of.hpp
Comment (hkaiser) [inline libs/core/type_support/include/hpx/type_support/detail/with_result_of.hpp:44]: Once we're in C++17 land we should replace this with a user defined deduction guideline to invoke CTAD.

File: libs/parallelism/execution/include/hpx/execution/algorithms/detach.hpp
Comment (hkaiser) [inline libs/parallelism/execution/include/hpx/execution/algorithms/detach.hpp:54]: This should be slightly less heavyweight:
```suggestion
            hpx::util::atomic_count count{0};
```

Comment (hkaiser) [reply libs/parallelism/execution/include/hpx/execution/algorithms/detach.hpp:54]: You might want to consider making the `count` private to avoid anybody starts messing with it. This would also require for the `intrusive_ptr_addref/release` functions to become friends of this class.

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (hkaiser) [inline libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:57]: Shouldn't this be:
```suggestion
                continuations.emplace_back(std::forward<F>(continuation));
```
?

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (hkaiser) [inline libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:70]: Could these two be combined into one `shared_ptr`?

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (hkaiser) [inline libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:92]: Same here, can those be combined?

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (hkaiser) [inline libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:196]: The is `;` superfluous here

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (hkaiser) [inline libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:208]:
```suggestion
                prev_state = std::move(state);
```

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (hkaiser) [inline libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:212]: Can the disjoint testing and setting of `prev_access` lead to races?

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (hkaiser) [inline libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:285]: Concurrent invocations of `start` may lead to unreported inconsistencies such that the `HPX_ASSERT` does not fire even if another thread is trying to execute `start` at the same time. I'm not sure if this is a real problem, though.

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (hkaiser) [inline libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:254]: Should the assignment operators be deleted as well?

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (hkaiser) [inline libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:469]: General question: all of the types defined here are move-only at best. Would using `unique_ptr` instead of `shared_ptr` be an option? And if yes, do we have to use a pointer type in the first place?

File: libs/core/type_support/include/hpx/type_support/detail/with_result_of.hpp
Comment (msimberg) [reply libs/core/type_support/include/hpx/type_support/detail/with_result_of.hpp:21]: Yeah, good catch. This was copied from those links.

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (msimberg) [reply libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:57]: Sort of. I wanted to constrain the signature to only take rvalue references, but this also works.

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (msimberg) [reply libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:70]: Not with the current setup unless I'm missing something. The value should live separately from the shared states (which act as triggers to continuations when the reference counts go to zero) and should be kept alive across all read-only and read-write accesses.

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (msimberg) [reply libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:92]: Same answer.

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (msimberg) [reply libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:212]: Yes and no, meaning it's intentionally not threadsafe. It's mentioned in the docstring. Not sure if it's worth making it threadsafe.

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (msimberg) [reply libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:285]: Can you explain how? `start` can only be called once on a (connected) sender.

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (msimberg) [reply libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:254]: Yep!

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (msimberg) [reply libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:469]: Read-only access requires the shared state to be a `shared_ptr`. Even in the read-write case, while only one sender sends the shared state in the wrapper, the following senders need to hold on to the same shared state to attach their continuations as well.

Comment (msimberg) [reply libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:469]: That said, I realize now that the value could be in a `unique_ptr` instead of a `shared_ptr`. The reference wrappers can hold references rather than `shared_ptr`s and the last continuation can release the `unique_ptr`. Now I'll have to reexamine if the shared states can do the same (I don't think so).

Comment (hkaiser) [reply libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:469]: If using a `unique_ptr<foo>` is possible, wouldn't be a plain `foo` possible as well? This should be viable at least for moveable or trivially copyable value types, shouldn't it?

Comment (msimberg) [reply libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:469]: Maybe you see something I don't. I want heap allocation to have the value live past the end of the `async_rw_mutex` destructor (I'd prefer not to block there).

Comment (hkaiser) [reply libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:469]: Don't pay attention to what I said, I was merely trying to understand the code base.

Comment (msimberg) [reply libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:469]: I will do exactly the opposite, thanks for looking through this!

File: libs/parallelism/execution/include/hpx/execution/algorithms/detach.hpp
Comment (msimberg) [reply libs/parallelism/execution/include/hpx/execution/algorithms/detach.hpp:54]: Note: I'll apply this in #5274. Is the main difference that `atomic_count` uses more relaxed memory orderings by default?

File: libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp
Comment (msimberg) [reply libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp:608]: I've updated it to use a `unique_ptr` for the guarded value now.

**Comment by msimberg** [comment]
I've removed the dependency on #5274, it was unnecessary.

**Comment by msimberg** [comment]
retest

**Comment by hkaiser** [comment]
@msimberg I still would like to come back to this, mostly to really understand how the facility is supposed to be used and what its semantics are. I still find it to be strange that two separate shared pointers are needed (four allocations!), for instance. My questions are: is the lifetime for those really such different, are those referred to from different places?.

**Comment by msimberg** [comment]
> @msimberg I still would like to come back to this, mostly to really understand how the facility is supposed to be used and what its semantics are.

No problem, I appreciate you looking through this!

> I still find it to be strange that two separate shared pointers are needed (four allocations!), for instance.

Two allocations?

> My questions are: is the lifetime for those really such different, are those referred to from different places?.

There's one allocation for each read-write access, and one shared by consecutive read-only accesses. You can think of the shared states as roughly forming a linked list (where a state refers to its successor through the callbacks which hold hold the actual instances of the successor shared state). Some sort of queue would likely also work for representing the same structure, but also needs heap allocations. A custom allocator is likely also a possibility, but I'd like avoid overcomplicating it at this stage at least.

It might be easiest to discuss this in more detail offline (in a call...).

**Review by biddisco — COMMENTED** [review]
I didn't look all the way through this yet, but would it not have been possible to use std::shared_mutex as the basic synchronization primitive - with try_lock giving exclusive access and try_shared_lock for shared access. A failure to acquire the lock being the trigger to swap from the current state to the next state. Just wondering out loud.

**Comment by msimberg** [comment]
> A failure to acquire the lock being the trigger to swap from the current state to the next state. Just wondering out loud.

Could you expand on that last part? How would you set things up in that case?

One thing I'm trying to avoid in this setup is to unnecessarily start tasks before they get access to the resource (and thus avoid unnecessary yielding). It's not entirely clear to me if that would happen with the shared mutex approach (in a naive implementation in my head it would do that).

I realize now that I may have left out a critical piece of information, which I hope helps clarify things. The senders should be triggered in the order that they are accessed, not just in any order as the shared mutex would allow. This is a requirement from the use case in DLA-Future where this essentially helps create an implicit DAG of accesses to a tile of a matrix. The name might be worth changing part of the confusion comes from this.

**Comment by msimberg** [comment]
@hkaiser the value is now moved from state to state instead of heap allocated. I'm pretty sure there are other optimizations that could be done here, but I'd like to leave those for the future if/when needed (please do point it out if you see something obviously bad though).

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

**Comment by hkaiser** [comment]
@msimberg thanks for pushing this forward!

**Pull request merged**
